nautobot/core/apps/__init__.py

```diff
@@ -0,0 +1,246 @@
+from django.apps import AppConfig
+from django.urls import reverse
+from operator import getitem
+from collections import OrderedDict
+
+from nautobot.extras.plugins.utils import import_object
+from nautobot.extras.registry import registry
+from nautobot.utilities.choices import ButtonActionColorChoices, ButtonActionIconChoices
+
+
+registry["nav_menu"] = {"tabs": {}}
+
+
+class NautobotConfig(AppConfig):
+    """
+    Custom AppConfig for Nautobot application.
+
+    Adds functionality to generate the HTML navigation menu using `navigation.py` files from nautbot
+    applications.
+    """
+
+    menu_tabs = "navigation.menu_tabs"
+
+    def ready(self):
+        """
+        Ready function initiates the import application.
+        """
+        menu_items = import_object(f"{self.name}.{self.menu_tabs}")
+        if menu_items is not None:
+            register_menu_items(menu_items)
+
+
+def register_menu_items(tab_list):
+    """
+    Using the imported object a dictionary is either created or updated with objects to create
+    the navbar.
+
+    The dictionary is built from four key objects, NavMenuTab, NavMenuGroup, NavMenuItem and
+    NavMenuButton. The Django template then uses this dictionary to generate the navbar HTML.
+    """
+    for nav_tab in tab_list:
+        if nav_tab.name not in registry["nav_menu"]["tabs"]:
+            registry["nav_menu"]["tabs"][nav_tab.name] = {
+                "weight": nav_tab.weight,
+                "groups": {},
+                "permissions": [],
+            }
+
+        tab_perms = []
```

Comment:
Should `tab_perms` and `group_perms` (as well as each of the `registry[...]["permissions"]`) be defined as sets rather than lists? So that we don't end up with redundant permissions being present in duplicate?